flink-ml-lib/src/main/java/org/apache/flink/ml/classification/logisticregression/LogisticRegressionWithFtrl.java

```diff
@@ -0,0 +1,380 @@
+                                o -> o.iterationId >= getMaxIter());
+        FTRL ftrl = new FTRL(getAlpha(), getBeta(), getReg(), getElasticNet());
+
+        DataStream<Tuple3<Long, Long, double[]>> rawModelData =
+                TrainingUtils.train(
+                        modelDim,
+                        trainData,
+                        ftrl,
+                        iterationStages,
+                        getNumServers(),
+                        getNumServerCores());
+
+        final long modelVersion = 0L;
+
+        DataStream<LogisticRegressionModelData> modelData =
+                rawModelData.map(
+                        tuple3 ->
+                                new LogisticRegressionModelData(
+                                        Vectors.dense(tuple3.f2),
+                                        tuple3.f0,
+                                        tuple3.f1,
+                                        modelVersion));
+
+        LogisticRegressionModel model =
+                new LogisticRegressionModel().setModelData(tEnv.fromDataStream(modelData));
+        ParamUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static LogisticRegressionWithFtrl load(StreamTableEnvironment tEnv, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+}
+
+/**
+ * An iteration stage that samples a batch of training data and computes the indices needed to
+ * compute gradients.
+ */
+class ComputeIndices extends ProcessStage<LogisticRegressionWithFtrlTrainingContext> {
+
+    @Override
+    public void process(LogisticRegressionWithFtrlTrainingContext context) throws Exception {
+        context.readInNextBatchData();
+        context.pullIndices = computeIndices(context.batchData);
+    }
+
+    public static long[] computeIndices(List<LabeledLargePointWithWeight> dataPoints) {
+        LongOpenHashSet indices = new LongOpenHashSet();
+        for (LabeledLargePointWithWeight dataPoint : dataPoints) {
+            long[] notZeros = dataPoint.features.f0;
+            for (long index : notZeros) {
+                indices.add(index);
+            }
+        }
+
+        long[] sortedIndices = new long[indices.size()];
+        Iterator<Long> iterator = indices.iterator();
+        int i = 0;
+        while (iterator.hasNext()) {
+            sortedIndices[i++] = iterator.next();
+        }
+        Arrays.sort(sortedIndices);
+        return sortedIndices;
+    }
+}
+
+/**
+ * An iteration stage that uses the pulled model values and sampled batch data to compute the
+ * gradients.
+ */
+class ComputeGradients extends ProcessStage<LogisticRegressionWithFtrlTrainingContext> {
```

Comment:
We can move these two classes outside `LogisticRegressionWithFtrl`. The problem is that `LabeledLargePointWithWeight` is universally applicable.

I have moved these two classes to `org.apache.flink.ml.common.ps.training` for now. We can get back to this discussion when we achieved consensus on `LabeledLargePointWithWeight`.

Reply:
I have removed `LabeledLargePointWithWeight` since we choose to generalize the implementation of `Vector`. This class have been moved to `org.apache.flink.ml.common.ps.training`